app/assets/javascripts/discourse/app/components/search-menu.js

```diff
@@ -0,0 +1,295 @@
+import Component from "@glimmer/component";
+import { inject as service } from "@ember/service";
+import { action } from "@ember/object";
+import { bind } from "discourse-common/utils/decorators";
+import { tracked } from "@glimmer/tracking";
+import {
+  isValidSearchTerm,
+  searchForTerm,
+  updateRecentSearches,
+} from "discourse/lib/search";
+import DiscourseURL from "discourse/lib/url";
+import discourseDebounce from "discourse-common/lib/debounce";
+import getURL from "discourse-common/lib/get-url";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import { search as searchCategoryTag } from "discourse/lib/category-tag-search";
+import userSearch from "discourse/lib/user-search";
+import { CANCELLED_STATUS } from "discourse/lib/autocomplete";
+import { cancel } from "@ember/runloop";
+
+const CATEGORY_SLUG_REGEXP = /(\#[a-zA-Z0-9\-:]*)$/gi;
+const USERNAME_REGEXP = /(\@[a-zA-Z0-9\-\_]*)$/gi;
+const SUGGESTIONS_REGEXP = /(in:|status:|order:|:)([a-zA-Z]*)$/gi;
+export const SEARCH_INPUT_ID = "search-term";
+export const SEARCH_BUTTON_ID = "search-button";
+export const MODIFIER_REGEXP = /.*(\#|\@|:).*$/gi;
+export const DEFAULT_TYPE_FILTER = "exclude_topics";
+
+export function focusSearchInput() {
+  document.getElementById(SEARCH_INPUT_ID).focus();
+}
+
+export function focusSearchButton() {
+  document.getElementById(SEARCH_BUTTON_ID).focus();
+}
+
+export default class SearchMenu extends Component {
+  @service search;
+  @service currentUser;
+  @service siteSettings;
+  @service appEvents;
+
+  @tracked inTopicContext = this.args.inTopicContext;
+  @tracked loading = false;
+  @tracked results = {};
+  @tracked noResults = false;
+  @tracked inPMInboxContext =
+    this.search.searchContext?.type === "private_messages";
+  @tracked typeFilter = DEFAULT_TYPE_FILTER;
+  @tracked suggestionKeyword = false;
+  @tracked suggestionResults = [];
+  @tracked invalidTerm = false;
+  _debouncer = null;
+  _activeSearch = null;
+
+  get includesTopics() {
+    return this.typeFilter !== DEFAULT_TYPE_FILTER;
+  }
+
+  get searchContext() {
+    if (this.inTopicContext || this.inPMInboxContext) {
+      return this.search.searchContext;
+    }
+
+    return false;
+  }
+
+  @bind
+  fullSearchUrl(opts) {
+    let url = "/search";
+    let params = "";
+
+    if (this.search.activeGlobalSearchTerm) {
+      params = `q=${this.search.activeGlobalSearchTerm}`;
+
+      if (this.searchContext?.type === "topic") {
+        params += ` topic:${this.searchContext.id}`;
+      } else if (this.searchContext?.type === "private_messages") {
+        params += " in:messages";
+      }
+    }
+
+    if (opts?.expanded) {
+      params += "&expanded=true";
+    }
+
+    if (params) {
+      params = new URLSearchParams(params);
+      url = `${url}?${params}`;
+    }
+
+    return getURL(url);
+  }
```

Comment:
We should use `URLSearchParams` to build the query string, so that it handles all the escaping. Right now we're building the string manually, and `URLSearchParams` is just parsing it and spitting it out again without changes.

```suggestion
    let params = new URLSearchParams();

    if (this.search.activeGlobalSearchTerm) {
      let q = this.search.activeGlobalSearchTerm;

      if (this.searchContext?.type === "topic") {
        q += ` topic:${this.searchContext.id}`;
      } else if (this.searchContext?.type === "private_messages") {
        q += " in:messages";
      }
      
      params.set("q", q);
    }

    if (opts?.expanded) {
      params.set("expanded", "true");
    }

    if (params.toString() !== "") {
      url = `${url}?${params}`;
    }

    return getURL(url);
  }
```